text/1023-splats.rst

```diff
@@ -0,0 +1,257 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2023-01-25
+    Authors: Yury Selivanov <yury@edgedb.com>
+
+
+==============================
+RFC 1023: Adding Splats Syntax
+==============================
+
+Abstract
+========
+
+We propose to add a splats syntax to EdgeQL. The closest equivalents to this
+proposal are fragments in GraphQL and `SELECT *` in SQL.
+
+Splats in ``select`` commands and in free object types will allow for better
+REPL user experience and will improve readability of complex queries and type
+expressions.
+
+
+Splats in select shapes
+=======================
+
+EdgeQL *shapes* is a syntax construct used to specify what exact properties,
+links, and ad-hoc computed data should be fetched by the query.
+
+This proposal assumes the following schema for all of its example code::
+
+    abstract type Person {
+      required property name -> str { constraint exclusive };
+    }
+
+    type Hero extending Person {
+      property secret_identity -> str;
+      multi link villains := .<nemesis[is Villain];
+    }
+
+    type Villain extending Person {
+      link nemesis -> Hero;
+    }
+
+With this schema in mind, here's an example of the EdgeQL's shape syntax::
+
+    select Person {
+      name,  # propety name
+
+      name_upper := str_upper(.name),  # computed property
+
+      villains: {  # fetch data traversing the "villains" link
+        name
+      },
+
+      [is Hero].secret_identity,  # fetch `secret_identity` property
+                                  # for all `Person` objects that are also
+                                  # instances of `Hero`
+    }
+
+We propose to enhance the shape construct with the following syntax:
+
+* ``*``: extend the shape with all properties (not links!) defined on the type,
```

Comment:
I like everything except that this only fetches properties. I understand why it looks conceptually nice. But I think for REPL use case it's vastly superior to have links too:
```
> SELECT Person {*}
{
  name: "John",
  villains: [
    Villain { id: 4d1fcc65-8aa7-11ed-ad95-2325e2521606 },
  ]
}
```
This is how links (objects) are fetched by default, you see ids (and types) of the objects.

Otherwise it looks like you still have to run describe every time to find out what `*` didn't show.

Reply:
I think I'm inclined to agree here.  The hesitation to include links in `*` is mostly dictated by performance worries, but then you are not really supposed to abuse `*` in application code, and for quick REPL queries and proofs-of-concept performance matters way less.

Reply:
This implies that we invert the "no shadowing" rule, because if `*` includes links, elaboration of them in subshapes becomes impossible.

Reply:
I still think that using `*` only for properties makes more sense for us for the following reasons:

* Users will inevitably use this in their application code (not just in REPL) and selecting links is expensive. Selecting all properties is typically a more common need than selecting all properties & all links. The performance would take a big hit if we select all links by default.

* We already have splats in JS (https://www.edgedb.com/docs/clients/js/querybuilder#select-objects, @colinhacks ❤️‍🩹) and the current API only expends `*` into properties.

* I have a feeling that when using splats in types (the second part of this RFC) it would be more valuable to unpack just properties.

That said I agree, that the REPL use case is important. So maybe we add `.**` syntax? @elprans suggests that potentially we can expand all links and all properties of all linked objects (1 level) to make it even more useful for exploration and messing around. Thoughts?

Reply:
Yes, double star is okay, I think, if it can also apply for standalone `User {**}`.

>  @elprans suggests that potentially we can expand all links and all properties of all linked objects (1 level) to make it even more useful for exploration and messing around. Thoughts?

Sounds plausible, but links are almost in all cases are cyclic. (i.e. computed `.<` is used too often, and we can't just ignore backlinks).

So I would like to have a way to have single level expansion and graph expansion as a separate operations, kinda:
1. `*` -- only properties
2. `*.*` -- properties and links
2. `*.**` -- links recursively, stopping when reaching the same type probably, i.e. `A -> B -> A` stops at the second `A`.

We can also probably invent some syntax for expanding fixed number of levels of links, like: `*.*.*` two level of links even if it's `User { friends: User }`.

But I'm not sure whether it's too much to learn.

Reply:
While we're at it. It we a syntax that differ like `*` with properties only, and `**` or `*.*` that includes a link, we want show some hint:
```
> SELECT {*}
{User { name: "John" }}
# Asterisk `*` only shows properties, use `**` to also show links.
```
Which means we want to add some "annotation" to the `CommandDataDescription` when bare star is used.

Reply:
I've updated the proposal with the `**` syntax.

I'm really not a fan of `*.*` and `*.**` variations. (FWIW @msullivan proposed something similar too). While they are certainly readable once you understand the mechanics, searching what they mean and how they work in the docs/web, as well as just quickly scanning through a query will be a pain for many users.

Reply:
A suggestion for the syntax:
- `*` for properties
- `*>` for properties and links
- chaining for specific depth, e.g. `*>*`, `*>*>`,`*>*>*`
- `***` for recursive

*(I'm new to EdgeQL, and I recognize this may not be good or viable for a variety of reasons.)*

Reply:
We discussed the possibility of adding syntax for deeper (arbitrary) expansions, but decided against it (at least for now), because:

- the syntax is ends up a bit arcane looking
- the real need for this is not entirely clear

@1st1, let's add a note to the RFC that this has been considered.